src/NServiceBus.Core/MessageMutators/MutatorRegistrationExtensions.cs

```diff
@@ -19,39 +19,36 @@ public static void RegisterMessageMutator(this EndpointConfiguration endpointCon
             Guard.AgainstNull(nameof(messageMutator), messageMutator);
 
             var registeredMutator = false;
-            foreach (var mutatorInterface in GetImplementedMutatorInterfaces(messageMutator))
-            {
-                // register the mutator with its specific mutator interface as it will be registered as object instead
-                endpointConfiguration.RegisterComponents(c => c.RegisterSingleton(mutatorInterface, messageMutator));
-                registeredMutator = true;
-            }
+           
+            var registry = endpointConfiguration.Settings.GetOrCreate<Mutators.RegisteredMutators>();
 
-            if (!registeredMutator)
+            if (messageMutator is IMutateIncomingMessages)
             {
-                throw new ArgumentException($"The given instance is not a valid message mutator. Implement one of the following mutator interfaces: {typeof(IMutateIncomingMessages).FullName}, {typeof(IMutateIncomingTransportMessages).FullName}, {typeof(IMutateOutgoingMessages).FullName} or {typeof(IMutateOutgoingTransportMessages).FullName}");
+                registry.IncomingMessage.Add((IMutateIncomingMessages)messageMutator);
+                registeredMutator = true;
             }
-        }
 
-        static IEnumerable<Type> GetImplementedMutatorInterfaces(object messageMutatorType)
-        {
-            if (messageMutatorType is IMutateIncomingMessages)
+            if (messageMutator is IMutateIncomingTransportMessages)
             {
-                yield return typeof(IMutateIncomingMessages);
+                registry.IncomingTransportMessage.Add((IMutateIncomingTransportMessages)messageMutator);
+                registeredMutator = true;
             }
 
-            if (messageMutatorType is IMutateIncomingTransportMessages)
+            if (messageMutator is IMutateOutgoingMessages)
```

Comment:
can't we use this syntax to avoid the cast on the `Add`: `if(messageMutator is IMutateOutgoingMessages outgoingMutator)...`?

Reply:
👍